### DISCUSSION TITLE
Automatic Timer for Pocketbase

### DISCUSSION THREAD

**Discussion by MuslemRahimi** [pocketbase/pocketbase#5080] [Q&A]
Hi,

I wanted to ask if it is possible to create a timer in pocketbase.
For instance, when a user is created I want to give him a free trial of 7 days. In my "user" collection I have "Tier" that can either take the value "Pro" or "Free".

**Comment by ganigeorgiev** [comment]
I'm not sure I follow. What happens after the free trial ends - delete the user?

There is no builtin "timer" functionality as part of the collections if that's what you are looking for. 
General timers and tickers can be created with the std Go [`time` package](https://pkg.go.dev/time) (_the JSVM doesn't have such options_).

But in your case I don't see why that would be needed. You can schedule a cron job with [Go](https://pocketbase.io/docs/go-jobs-scheduling/) or [JSVM](https://pocketbase.io/docs/js-jobs-scheduling/) that will query the users matching your requirements and then update or delete the specific records. For example (_I haven't tested it_):

- Go:
    ```go
    scheduler := cron.New()

    // eg. check twice a day (or more often if you want to be more precise)
    scheduler.MustAdd("free_trial", "0 */12 * * *", func() {
        // with the next release you'll be able to use directly
        // types.NowDateTime().AddDate(0, 0, 7)
        minDate, _ := types.ParseDateTime(time.Now().AddDate(0, 0, 7)) 

        records, err := app.Dao().FindRecordsByExpr("users",
            dbx.NewExp("tier = 'Free' AND created >= {:minDate}", dbx.Params{"minDate": minDate}),
        )
        if err != nil {
            app.Logger().Error("Failed to fetch free trial users", "error", err)
            return
        }

        for _, r := range records {
            // delete the free trial user?
            if err := app.Dao().DeleteRecord(r); err != nil {
                app.Logger().Error("Failed to delete free trial user", "error", err, "id", r.Id)
            }
        }
    })

    scheduler.Start()
    ```
    
- JSVM:
    ```js
    // eg. check twice a day (or more often if you want to be more precise)
    cronAdd("free_trial", "0 */12 * * *", () => {
        const date = new Date();
        date.setDate(date.getDate() + 7);

        try {
            const records = $app.dao().findRecordsByExpr("users",
                $dbx.exp("tier = 'Free' AND created >= {:minDate}", { "minDate": date.toISOString().replace("T", " ") }),
            )

            for (let r of records) {
                // delete the free trial user?
                try {
                    $app.dao().deleteRecord(r)
                } catch (err) {
                    $app.logger().error("Failed to delete free trial user", "error", err, "id", r.id)
                }
            }
        } catch (err) {
            $app.logger().error("Failed to fetch free trial users", "error", err)
        }
    })
    ```

**Reply by MuslemRahimi** [reply]
Thank you for the reply.

I thought that maybe this could be done more elegant in pocketbase as a feature. I wanted to give users a free trial to the "Pro" tier for 7 days and automatically downgrade them to the "Free" Tier afterwards.

**Reply by ganigeorgiev** [reply]
> I thought that maybe this could be done more elegant in pocketbase as a feature.

No, I don't think it can be generalized as it would depend too much on the specific application flow.
